Framework/PythonInterface/plugins/algorithms/CreateGroupingByComponent.py

```diff
@@ -0,0 +1,184 @@
+        group_string = ",".join([self.get_component_group_string(p, num_divisions) for p in direct_parents])
+
+        # print this string in debug mode
+        logger.debug(group_string)
+
+        # use this string as input for CreateGroupingWorkspace
+        group_ws, *_ = self.exec_child_alg(
+            "CreateGroupingWorkspace",
+            InstrumentName=instr,
+            ComponentName=self.info.name(self.info.root()),
+            CustomGroupingString=group_string,
+            OutputWorkspace="__group_ws",
+        )
+
+        self.setProperty("OutputWorkspace", group_ws)
+        ADS.remove("__instr_tmp_ws")
+        ADS.remove("__group_ws")
+
+    def exec_child_alg(self, alg_name: str, **kwargs):
+        alg = self.createChildAlgorithm(alg_name, enableLogging=False)
+        alg.setAlwaysStoreInADS(False)
+        alg.initialize()
+        alg.setProperties(kwargs)
+        alg.execute()
+        out_props = tuple(alg.getProperty(prop).value for prop in alg.outputProperties())
+        return out_props[0] if len(out_props) == 1 else out_props
+
+    def get_direct_parents_of_targets(self, root_idx: int):
+        """
+        find all the parent nodes in the instrument tree with at least one child node which meets the search criteria
+        """
+        parents = set()
+        stack = [int(root_idx)]
+        while stack:
+            idx = stack.pop()
+            if self._branch_should_be_excluded(idx):
+                continue
+            children = [int(c) for c in self.info.children(idx)]
+            # if any child is a target, this node is a direct parent
+            if any(self.is_target(c) for c in children):
+                parents.add(idx)
+            stack.extend(c for c in children if not self.is_target(c))
+        return sorted(parents)
+
+    def get_component_group_string(self, parent, n=1):
+        """
+        for a given parent node, find all the children which are detectors, and split them into n groups - formatted as
+        a custom detector grouping string for CreateGroupingWorkspace
+        """
+        all_sub_arrays = [
+            arr
+            for arr in np.array_split([str(self.det_id_dict[det_ind]) for det_ind in self.info.detectorsInSubtree(int(parent))], n)
+            if len(arr) > 0
+        ]
+        return ",".join(["+".join(sub_array) for sub_array in all_sub_arrays])
+
+    def _component_name_contains(self, idx, target):
+        return target in self.info.name(int(idx))
+
+    def _is_target_include(self, idx):
+        return self._component_name_contains(idx, self.target)
+
+    def _is_target_exclude(self, idx):
+        return self._is_target_include(idx) and not self._component_name_contains(idx, self.null)
+
+    def _branch_should_be_excluded(self, idx):
+        if not self.exclude_branches:
+            return False
+        return np.any([self._component_name_contains(idx, branch) for branch in self.exclude_branches])
```

Comment:
```suggestion
    def _branch_should_be_excluded(self, idx):
        return np.any([self._component_name_contains(idx, branch) for branch in self.exclude_branches])
```
if `self.exclude_branches==[]` then this will return False (well `np.False_`)